File: problems/java/VendingMachineDemo/src/states/DispenseState.java
package states;

import models.Product;
import models.Rack;
import services.VendingMachine;

public class DispenseState implements State {

    @Override
    public void insertMoney(VendingMachine vendingMachine, Double amount) {
        System.out.println("[DispenseState] Can't insert money while dispensing");
    }

    @Override
    public void selectProduct(VendingMachine vendingMachine, Integer rackNumber) {
        System.out.println("[DispenseState] Can't select product while dispensing");
    }

    public void dispense(VendingMachine vendingMachine, Rack rack) {
        Product product = rack.getProduct();
        System.out.println("[DispenseState] Dispensing: " + product.getName());
        rack.dispenseOne();
        double paid = vendingMachine.getCurrentAmount();
        double change = paid - product.getPrice();
        if (change > 0) {
            System.out.println("[DispenseState] Change: Rs " + change);
        }
    }
}

File: problems/python/VendingMachineDemo/src/states/dispense_state.py
"""Dispense state: delivers product and returns change."""
from states.state import State
from models.rack import Rack


class DispenseState(State):
    def insert_money(self, vending_machine: "VendingMachine", amount: float) -> None:
        print("[DispenseState] Can't insert money while dispensing")

    def select_product(self, vending_machine: "VendingMachine", rack_number: int) -> None:
        print("[DispenseState] Can't select product while dispensing")

    def dispense(self, vending_machine: "VendingMachine", rack: Rack) -> None:
        product = rack.get_product()
        print(f"[DispenseState] Dispensing: {product.name}")
        rack.dispense_one()
        paid = vending_machine.get_current_amount()
        change = paid - product.price
        if change > 0:
            print(f"[DispenseState] Change: Rs {change}")
